keycloak/providers/authenticator/hmda/src/main/java/gov/cfpb/keycloak/authenticator/hmda/HmdaValidInstitutionsFormAction.java

```diff
@@ -0,0 +1,268 @@
+        List<FormMessage> errors = new ArrayList<>();
+
+        logger.info("Form Data: " + formData);
+
+        String instFieldVal = formData.getFirst(FIELD_INSTITUTIONS);
+        if (Validation.isBlank(instFieldVal)) {
+            // FIXME: Make these errors consistent with Keycloak's own errors
+            errors.add(new FormMessage(FIELD_INSTITUTIONS, "Institution(s) not set"));
+            context.validationError(formData, errors);
+            context.error(Errors.INVALID_REGISTRATION);
+            return;
+        }
+
+        try {
+            String[] instIds = instFieldVal.split(",");
+            String email = formData.getFirst(RegistrationPage.FIELD_EMAIL);
+            domain = email.split("@")[1];
+
+            logger.info("Email Domain: " + domain);
+
+
+            Set<Institution> domainInsts = findInstitutionsByDomain(domain);
+
+            if (domainInsts.isEmpty()) {
+                errors.add(new FormMessage(RegistrationPage.FIELD_EMAIL, "Email domain '" + domain + "' not in CFPB whitelist"));
+                context.validationError(formData, errors);
+                context.error(Errors.INVALID_REGISTRATION);
+                return;
+            }
+
+            for (String instId : instIds) {
+
+                Institution inst = getInstitution(instId);
+                ;
+
+                if (inst == null) {
+                    errors.add(new FormMessage(FIELD_INSTITUTIONS, "No institution found with ID " + instId));
+                }
+
+                if (!domainInsts.contains(inst)) {
+                    errors.add(new FormMessage(
+                            FIELD_INSTITUTIONS,
+                            "Institution \"" + inst.getName() + "\" not associated with email domain \"" + domain + "\""));
+                }
+            }
+
+            if (!errors.isEmpty()) {
+                context.validationError(formData, errors);
+                context.error(Errors.INVALID_REGISTRATION);
+                return;
+            }
+
+            context.success();
+        } catch (Exception e) {
+            String message = "Error occurred while validating institution(s) against \"" + domain + "\" domain";
+            logger.error(message, e);
+            errors.add(new FormMessage(FIELD_INSTITUTIONS, message));
+            context.validationError(formData, errors);
+            context.error(Errors.INVALID_REGISTRATION);
+        }
+    }
+
+    @Override
+    public void success(FormContext form) {
+
+    }
+
+    @Override
+    public boolean requiresUser() {
+        return false;
+    }
+
+    @Override
+    public boolean configuredFor(KeycloakSession session, RealmModel realm, UserModel user) {
+        // FIXME: We could check to make sure it is the "hmda" realm?
+        return true;
+    }
+
+    @Override
+    public void setRequiredActions(KeycloakSession session, RealmModel realm, UserModel user) {
+        // FIXME: Anything we should do with this?
```

Comment:
Seems like the verify email required action is inherited, which is fine for now. This is where we'd put account recovery questions, etc. but that can come in a later PR.